### ISSUE TITLE
Runtime API does not update default / fallback crt

### ISSUE THREAD

**Issue by jcmoraisjr** [haproxy/haproxy#1143]
## Detailed description of the problem

The runtime API does not update the default / fallback entry of a crt-list.

## Expected behavior

If updating a certificate used as default / fallback via runtime API, domains that doesn't match any of the sni in the crt-list should use the new certificate.

## Steps to reproduce the behavior

haproxy.cfg

```
global
  stats socket /tmp/haproxy.sock level admin
  stats timeout 1m
defaults
  timeout client 1m
  timeout server 1m
  timeout connect 5s
listen l
  mode http
  bind :8443 ssl crt-list crt.list
  http-request return
```

crt.list

```
default.pem !*
default.pem dom1.local
dom2.pem dom2.local
```

Starting state:

```
$ openssl x509 -noout -dates -subject -in default.pem
notBefore=Feb 20 13:47:58 2021 GMT
notAfter=Mar 22 13:47:58 2021 GMT
subject= /CN=*.local

$ openssl x509 -noout -dates -subject -in dom2.pem
notBefore=Feb 20 13:48:41 2021 GMT
notAfter=Mar 22 13:48:41 2021 GMT
subject= /CN=dom2.local
```

Test one - dates match the current state:

```
$ : |openssl s_client -connect localhost:8443 -servername any.local 2>/dev/null |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:47:58 2021 GMT
notAfter=Mar 22 13:47:58 2021 GMT
subject= /CN=*.local

$ : |openssl s_client -connect localhost:8443 -servername dom1.local 2>/dev/null |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:47:58 2021 GMT
notAfter=Mar 22 13:47:58 2021 GMT
subject= /CN=*.local

$ : |openssl s_client -connect localhost:8443 -servername dom2.local 2>/dev/null |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:48:41 2021 GMT
notAfter=Mar 22 13:48:41 2021 GMT
subject= /CN=dom2.local
```

New state - I reissued the default and dom2 certificates:

```
$ openssl x509 -noout -dates -subject -in default.pem
notBefore=Feb 20 13:53:30 2021 GMT
notAfter=Mar 22 13:53:30 2021 GMT
subject= /CN=*.local

$ openssl x509 -noout -dates -subject -in dom2.pem
notBefore=Feb 20 13:53:44 2021 GMT
notAfter=Mar 22 13:53:44 2021 GMT
subject= /CN=dom2.local
```

Update only dom2.pem using the runtime API:

```
$ echo -e "set ssl cert dom2.pem <<\n$(cat dom2.pem)\n" |socat - /tmp/haproxy.sock
Transaction created for certificate dom2.pem!

$ echo -e "commit ssl cert dom2.pem" |socat - /tmp/haproxy.sock
Committing dom2.pem.
Success!
```

Test two - dates also match the current state:

```
$ : |openssl s_client -connect localhost:8443 -servername any.local 2>/dev/null |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:47:58 2021 GMT
notAfter=Mar 22 13:47:58 2021 GMT
subject= /CN=*.local

$ : |openssl s_client -connect localhost:8443 -servername dom1.local 2>/dev/null |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:47:58 2021 GMT
notAfter=Mar 22 13:47:58 2021 GMT
subject= /CN=*.local

$ : |openssl s_client -connect localhost:8443 -servername dom2.local 2>/dev/null |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:53:44 2021 GMT
notAfter=Mar 22 13:53:44 2021 GMT
subject= /CN=dom2.local
```

Now update the default.pem using the runtime API:

```
$ echo -e "set ssl cert default.pem <<\n$(cat default.pem)\n" |socat - /tmp/haproxy.sock
Transaction created for certificate default.pem!

$ echo -e "commit ssl cert default.pem" |socat - /tmp/haproxy.sock
Committing default.pem..
Success!
```

Test three - dates match all the entries but the default/fallback:

```
$ : |openssl s_client -connect localhost:8443 -servername dom1.local |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:53:30 2021 GMT
notAfter=Mar 22 13:53:30 2021 GMT
subject= /CN=*.local

$ : |openssl s_client -connect localhost:8443 -servername any.local |openssl x509 -noout -dates -subject
notBefore=Feb 20 13:47:58 2021 GMT
notAfter=Mar 22 13:47:58 2021 GMT
subject= /CN=*.local
```

## Do you have any idea what may have caused this?

No

## Do you have an idea how to solve the issue?

No

## What is your configuration?

Already in the steps to reproduce the behavior.

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.3.4-10189c9 2021/01/13 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.4.html
Running on: Darwin 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64
Build options :
  TARGET  = generic
  CPU     = generic
  CC      = clang
  CFLAGS  =
  OPTIONS = USE_KQUEUE=1 USE_PCRE=1 USE_POLL=1 USE_THREAD=1 USE_OPENSSL=1 USE_ZLIB=1
  DEBUG   =

Feature list : -EPOLL +KQUEUE -NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY -LINUX_TPROXY -LINUX_SPLICE -LIBCRYPT -CRYPT_H -GETADDRINFO +OPENSSL -LUA -FUTEX -ACCEPT4 -CLOSEFROM +ZLIB -SLZ -CPU_AFFINITY -TFO -NS -DL -RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER -PRCTL -THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 1.1.1i  8 Dec 2020
Running on OpenSSL version : OpenSSL 1.1.1i  8 Dec 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using:
Built with PCRE version : 8.44 2020-02-12
Running on PCRE version : 8.44 2020-02-12
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): no
Built with clang compiler version 12.0.0 (clang-1200.0.32.28)

Available polling systems :
     kqueue : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use kqueue.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by wlallemand** [comment]
Thanks for the report, I took a brief look at the code and it looks like the update of the default SSL_CTX is missing, fortunately we have the is_default boolean in the ckch instance and the architecture was made for this so it shouldn't be difficult to fix. I guess I missed this part when merging the feature. Also, it's not related to crt-list but to the default certificate in any case.

**Comment by wlallemand** [comment]
This was fixed and backported as far as 2.2.

**Comment by lukastribus** [comment]
Jarno bisected a startup crash in 2.2.12 to this fix:
https://www.mail-archive.com/haproxy@formilux.org/msg40136.html

**Comment by wlallemand** [comment]
That's a regression which is 2.2 only, fixed in 2.2.13. Thanks!
